hook.c

```diff
@@ -31,6 +31,14 @@ ATTR_NONNULL_ALL void hook_address(uint8_t type, struct in_addr* address, uint8_
     action = (char*)"release";
     break;
 
+  case HOOK_INFORM:
+    action = (char*)"inform";
+    break;
+
+  case HOOK_BEFORE_LEASE:
+    action = (char*)"beforelease";
+    break;
+
```

Comment:
Good catch on the inform hook. About the before_lease hook I am a bit unsure. What would be the use case, given we have a lease hook just a few lines down?

Reply:
it happens before the packet is send, so the usecase would be to delay sending that packet out until the hook is finished (not sure if the hook_address function is blocking, the goal would be that it is)

currently in the graz free wifi, internet doesn't work immediately but rather it takes until the route propagates up to the gateways (usually arround 5s). some devices don't check twice and just use mobile data or switch the wifi, which is bad ux.

the idea is to use the hook to add the address to l3roamd (sets route which starts propagation), run a ping with the source address of the client (this is going to be a custom tool, since we only assign a route and not an address for the client to the roam interface) and wait until either that works or 5 seconds pass (to avoid dhcp timeout)

this is the best solution I came up with so far, but we could theoretically also have that be handeled by ddhcpd itself or some other way

Reply:
hrm... so beforelease is nonblocking and also I've got another solution: add hook (claim) that is called like the lease hook, which, for ex l3roamd can use to add the address to the network

Reply:
I like that solution (adding the address to the network). This saves a discovery cycle which depends on usage of the address. Either way, those hooks are called non-blocking.
It is difficult to identify whether a route has fully propagated especially since the idea is to do this on demand only (to keep the route tables as small as possible).

Reply:
I have a patch for the claim hook for ffgraz ready. It runs the claim hook for all addresses that are free and have not yet had the claim hook run. When an address is leased out it changes the hook state back to zero, so if it becomes free the hook is ran again.

Instead of pushing all the addresses into a fake client on l3roamd, the script that gets triggered by the hook simply adds the route. (l3roamd clients eventually switch to inactive)

With our current settings that's 4 addresses per node (we have 2 bits as block size) and we'll approximate 50 gluon nodes this year max, so 500 route entries, which is ok for now.

I could extend it so it only adds two or even one address ahead of time, since propagation is only a few seconds.